### PR TITLE
TeraPool merge request preview

### PR DESCRIPTION
**Pull request by yichao-zh**
Summarize the changes concisely.

## Changelog
- RTL changes:
-Tile/Group/Cluster levels now could make by config for both TeraPool and MemPool;
-Add new Subgroup RTL;
- Testbench changes for TeraPool's hierarchy name, and some changes for post-layout simulation.
- QuestaSim waveform scripts update for TeraPool.

### Added
- Subgroup RTL file
### Changed
- Tile/Group/Cluster level RTL files.
- mempool.tb
- Bender.yml
- hardware/Makefile
- config.mk and terapool.mk

### Fixed

(Reference to issues, labels, and related merge requests)

## Checklist

- [ ] Automated tests pass
- [ ] Changelog updated
- [ ] Code style guideline is observed

Please check our [contributing guidelines](CONTRIBUTING.md) before opening a Pull Request.

### CODE REVIEW

**Review by mbertuletti — CHANGES_REQUESTED** [review]
Did you try vcs simulation? The simulation scripts should be modified for vcs too.

File: Bender.yml [hunk 1]
```diff
@@ -3,7 +3,7 @@
 # SPDX-License-Identifier: Apache-2.0
 
 package:
-  name: mempool
+  name: "terapool"
```
Comment (SamuelRiedel) [inline Bender.yml:6]: Why did you rename the Bender module?

File: config/config.mk [hunk 2]
```diff
@@ -32,6 +32,9 @@ boot_addr ?= 2684354560 # A0000000
 l2_base  ?= 2147483648 # 80000000
 l2_size  ?= 4194304    # 400000
 l2_banks ?= 4
+ifeq ($(config), terapool)
+  l2_banks = 16
+endif
```
Comment (SamuelRiedel) [inline config/config.mk:37]: This should not be necessary if you have set the l2 banks in `terapool.mk` already. The specific config will override the default from this file.

File: config/config.mk [hunk 3]
```diff
@@ -51,6 +54,9 @@ ro_line_width ?= 512
 
 # Number of DMA backends in each group
 dmas_per_group ?= 4
+ifeq ($(config), terapool)
+	dmas_per_group = 8
+endif
```
Comment (SamuelRiedel) [inline config/config.mk:59]: The same applies here.

File: software/apps/Makefile [hunk 4]
```diff
@@ -24,9 +24,14 @@ else
 	ALL := $(filter-out systolic/%,$(APPS))
 endif
 
+ALL_LLVM := $(filter-out synth, $(ALL))
+
```
Comment (SamuelRiedel) [inline software/apps/Makefile:28]: This is because we don't have the extension in LLVM yet, right? Once we fix this, we can remove this distinction again?

File: software/apps/synth/main.c [hunk 5]
```diff
@@ -13,27 +14,89 @@
...
   }
+#endif
 
   // wait until all cores have finished
   mempool_barrier(num_cores);
 
-  return error;
-}
+  return 0;
+}
```
Comment (SamuelRiedel) [inline software/apps/synth/main.c:102]: Missing EOF

File: software/apps/synth/main.c [hunk 6]
```diff
@@ -13,27 +14,89 @@
...
+// Macros
+#define UNROLL (4)
+#define N (NUM_CORES * UNROLL)
+#define TARGET_MEMPOOL (1)
+#define TARGET_TERAPOOL (0)
 
-int32_t block_a[N] __attribute__((section(".l1_prio")));
-int volatile error __attribute__((section(".l1")));
+// Globals
+int32_t block_a[N] __attribute__((aligned(NUM_CORES * 4), section(".l1")));
```
Comment (SamuelRiedel) [inline software/apps/synth/main.c:24]: Where is this four coming from? Do you want to align to the L1 start so all accesses are local?

File: software/apps/synth/main.c [hunk 7]
```diff
@@ -13,27 +14,89 @@
 #include "runtime.h"
 #include "synchronization.h"
 
-#define N 16
+// Macros
+#define UNROLL (4)
+#define N (NUM_CORES * UNROLL)
```
Comment (SamuelRiedel) [inline software/apps/synth/main.c:19]: Is `N` really `(NUM_CORES * UNROLL)` or should it be `(NUM_CORES * BANKING_FACTOR)`?

File: software/runtime/kernel/synth.h [hunk 8]
```diff
@@ -142,13 +461,340 @@ static inline void power_profile(int32_t *const a) {
...
+  register int32_t a0;
+  register int32_t a1;
+  register int32_t a2;
+  register int32_t a3;
+  register int32_t a4;
+  register int32_t a5;
+  register int32_t a6;
+  register int32_t a7;
+  __asm__ volatile(
+      // NOP to wait for instruction cache refill
```
Comment (SamuelRiedel) [inline software/runtime/kernel/synth.h:533]: Why do we have to do so many nops? Won't they pollute the I$, and we should just execute the loop multiple times?

File: hardware/src/mempool_group.sv [hunk 9]
```diff
@@ -75,578 +97,1067 @@ module mempool_group
...
-      .tcdm_dma_resp_valid_o   (tcdm_dma_resp_valid[t]                         ),
-      .tcdm_dma_resp_ready_i   (tcdm_dma_resp_ready[t]                         ),
-      // AXI interface
-      .axi_mst_req_o           (axi_tile_req[t]                                ),
-      .axi_mst_resp_i          (axi_tile_resp[t]                               ),
-      // Wake up interface
-      .wake_up_i               (wake_up_q[t*NumCoresPerTile +: NumCoresPerTile])
+  `ifdef TERAPOOL
+  /*********************
+   *  TeraPool Section *
```
Comment (SamuelRiedel) [inline hardware/src/mempool_group.sv:115]: What happens if we have one subgroup per group? Would we then have MemPool?

File: hardware/Makefile [hunk 10]
```diff
@@ -88,14 +88,22 @@ vlog_args += -suppress vlog-2583 -suppress vlog-13314 -suppress vlog-13233
...
-vlog_defs += -DBOOT_ADDR=$(boot_addr) -DXPULPIMG=$(xpulpimg)
+vlog_defs += -DL2_BASE=32\'d$(l2_base) -DL2_SIZE=32\'d$(l2_size) -DL2_BANKS=$(l2_banks)
+vlog_defs += -DBOOT_ADDR=32\'d$(boot_addr) -DXPULPIMG=$(xpulpimg)
 vlog_defs += -DSNITCH_TRACE=$(snitch_trace)
 vlog_defs += -DAXI_DATA_WIDTH=$(axi_data_width)
 vlog_defs += -DRO_LINE_WIDTH=$(ro_line_width)
 vlog_defs += -DDMAS_PER_GROUP=$(dmas_per_group)
 vlog_defs += -DAXI_HIER_RADIX=$(axi_hier_radix) -DAXI_MASTERS_PER_GROUP=$(axi_masters_per_group)
 vlog_defs += -DSEQ_MEM_SIZE=$(seq_mem_size) -DXQUEUE_SIZE=$(xqueue_size)
+ifdef terapool
```
Comment (SamuelRiedel) [inline hardware/Makefile:99]: We should define all defines for all systems if possible and have as few ifdefs as possible. So you can actually just remove this ifdef.

File: hardware/src/mempool_pkg.sv [hunk 11]
```diff
@@ -286,4 +286,38 @@ package mempool_pkg;
   // Replaces core with a traffic generator
   parameter bit TrafficGeneration  = `ifdef TRAFFIC_GEN `TRAFFIC_GEN `else 0 `endif;
 
+  /*************************
+   *  TeraPool PARAMETERS  *
+   *************************/
+
+  localparam integer unsigned NumSubGroupsPerGroup = `ifdef TERAPOOL `NUM_SUB_GROUPS_PER_GROUP `else 1 `endif;
+
+  `ifdef TERAPOOL
```
Comment (SamuelRiedel) [inline hardware/src/mempool_pkg.sv:295]: Here, we don't need the ifdef, right? We could make a comment that this is only used by TeraPool, but we can always define it. Or am I mistaken?

File: hardware/src/mempool_tile.sv [hunk 12]
```diff
@@ -410,13 +420,30 @@ module mempool_tile
...
+    tc_clk_gating i_ckg (
+      .clk_i    (clk_i                     ),
+      .test_en_i(1'b0                      ),
+      .en_i     (req_valid || sram_active_q),
+      .clk_o    (sram_clk                  )
+    );
+`else
+    assign sram_clk = clk_i;
+`endif
+
```
Comment (SamuelRiedel) [inline hardware/src/mempool_tile.sv:439]: This can be done a bit cleaner in the technology itself. We can discuss this offline.

File: hardware/src/mempool_tile.sv [hunk 13]
```diff
@@ -665,30 +699,84 @@ module mempool_tile
...
-         prescramble_tcdm_req_tgt_addr[ByteOffset +: idx_width(NumBanksPerTile)],                                                                              // Bank
-         prescramble_tcdm_req_tgt_addr[ByteOffset + idx_width(NumBanksPerTile) +: $clog2(NumTilesPerGroup)]}); // Tile
-    end
-    if (NumGroups == 1) begin : gen_remote_req_interco_tgt_sel
-      assign remote_req_interco_tgt_sel[c] = 1'b0;
-    end else begin : gen_remote_req_interco_tgt_sel
-      // Output port depends on both the target and initiator group
-      assign remote_req_interco_tgt_sel[c] = (prescramble_tcdm_req_tgt_addr[ByteOffset + $clog2(NumBanksPerTile) + $clog2(NumTilesPerGroup) +: $clog2(NumGroups)]) ^ group_id;
-    end
+    `ifdef TERAPOOL
```
Comment (SamuelRiedel) [inline hardware/src/mempool_tile.sv:702]: This ifdef could be a regular generate statement, right?

File: hardware/src/mempool_cluster.sv [hunk 14]
```diff
@@ -125,82 +125,355 @@ module mempool_cluster
...
-        assign tcdm_slave_req_valid[tgt][ini ^ tgt]  = tcdm_master_req_valid[ini][ini ^ tgt];
-        assign tcdm_master_req_ready[ini][ini ^ tgt] = tcdm_slave_req_ready[tgt][ini ^ tgt];
-
-        assign tcdm_master_resp[tgt][ini ^ tgt]       = tcdm_slave_resp[ini][ini ^ tgt];
-        assign tcdm_master_resp_valid[tgt][ini ^ tgt] = tcdm_slave_resp_valid[ini][ini ^ tgt];
-        assign tcdm_slave_resp_ready[ini][ini ^ tgt]  = tcdm_master_resp_ready[tgt][ini ^ tgt];
-      end: gen_remote_interconnections
-    end: gen_interconnections_tgt
-  end: gen_interconnections_ini
+  `ifdef TERAPOOL
```
Comment (SamuelRiedel) [inline hardware/src/mempool_cluster.sv:128]: This could also be a conditional generate statement. Also, is the only difference between MemPool and TeraPool here the registers? Because it looks like the code is pretty much the same, but we have registers in between. If so, we could just conditionally instantiate the registers and share the code. Of course, the other difference is the post-layout, but that could also be shared.

**Review by SamuelRiedel — COMMENTED** [review]
Thanks a lot Yichao. It looks already very good to me. I have a few comments and general questions.

- What prevents us from using generate statements instead of ifdefs when instantiating tera/mempool things?
- I see that the group is quite different, but for the cluster and the tile, could mempool just be the special case where we only have one subgroup per group? And then have the same instantiation of the modules and less code duplications?

File: software/apps/Makefile (hunk 4 above)
Comment (yichao-zh) [reply software/apps/Makefile:28]: Yes, you are right, I add these just for the CI checking on ASM codes.

File: hardware/src/mempool_group.sv [hunk 15]
```diff
@@ -13,7 +13,9 @@ module mempool_group
   // TCDM
   parameter addr_t       TCDMBaseAddr = 32'b0,
   // Boot address
-  parameter logic [31:0] BootAddr     = 32'h0000_1000
+  parameter logic [31:0] BootAddr     = 32'h0000_1000,
+  // For post-synthesis
+  parameter int unsigned GroupId      = 32'd0
```
[the remark this answers is not in the document]
Comment (yichao-zh) [reply hardware/src/mempool_group.sv:18]: These `IDs` are used for the post-synth/layout simulation. In the TeraPool case, the simulation runtime is too slow, so I just replace one of the Groups/Subgroups (e.g. Id=0) with a post-layout netlist to speed up the simulation runtime.

File: hardware/src/mempool_tile.sv (hunk 12 above)
Comment (yichao-zh) [reply hardware/src/mempool_tile.sv:439]: Ok, I see your point, sure if you have a solution of this, we can replace these lines. Let's discuss this offline when you have time.

File: hardware/src/mempool_group.sv (hunk 9 above)
Comment (yichao-zh) [reply hardware/src/mempool_group.sv:115]: To be honest, I thought about this style at the beginning. However, it will add one more hierarchy `subgroup` into the MemPool. Then as the hierarchical name structure changed for MemPool, it may cause many scripts to fail, for example, ModelSim wave.tcl and post-layout simulation, power simulation, and backend flows.  If we want to go this way, we may need a deep change.

File: hardware/src/mempool_group.sv (hunk 15 above)
[the remark this answers is not in the document]
Comment (SamuelRiedel) [reply hardware/src/mempool_group.sv:18]: So you need the ID of the group to know whether you are in the group where you will replace the subgroup with the netlist? In that case, couldn't we name the parameter something like netlist/postlayout directly to make it more clear that this is just to instantiate the netlist and not to define the ID itself?

File: hardware/src/mempool_group.sv (hunk 9 above)
Comment (SamuelRiedel) [reply hardware/src/mempool_group.sv:115]: it might be worth putting in the effort to keep the code much cleaner and more maintainable. But we can also proceed like this for now and then see whether we can converge the two designs later.

File: hardware/Makefile (hunk 10 above)
Comment (yichao-zh) [reply hardware/Makefile:99]: Hi Sam, the only thing left for now should be these `ifdef` stuff, if you agree, we can keep it temporary for this time, then we can discuss how to make it better when we have time in the future.

Comment (SamuelRiedel) [reply hardware/Makefile:99]: We can leave them in the RTL, but we could remove them in the Makefiles, right? Then, we are sure to have all the parameters in all configurations, even if some don't do anything. It is better to have them at a reasonable default than not defined for MemPool.

File: hardware/src/mempool_group.sv (hunk 15 above)
[the remark this answers is not in the document]
Comment (yichao-zh) [reply hardware/src/mempool_group.sv:18]: Dear Sam, Thanks for your suggestion, but I still don't have a clear idea of how I can name the parameter like "netlist/postlayout". Could you give me a toy example? In this case, I only replaced the subgroup design in Group 0, so I need the Group_ID to determine which is Group 0. I am sorry if I misunderstood something.

File: hardware/Makefile (hunk 10 above)
Comment (yichao-zh) [reply hardware/Makefile:99]: Additionally, I've used `ifdef TERAPOOL` in all the places where I made changes for the TeraPool RTL. If I set `vlog_defs += -DTERAPOOL=$(terapool)` in the `Makefile` for every configuration, it still presents an issue. This is because `TERAPOOL` will always be defined, even in `MemPool/MinPool` configurations, but with an "Empty" value. 
In this case, I think use the condition statement in Makefile is most easy way for it:
```
ifdef terapool
  vlog_defs += -DTERAPOOL=$(terapool)
 ```
 Any further suggestions?

Comment (SamuelRiedel) [reply hardware/Makefile:99]: Hey Yichao,

About the first aspect, yes, that is exactly what I had in mind as well. But instead of defining it in every configuration file, we just define a default in the `config/config.mk` file. We can also add a comment along the lines of: "This parameter is only used for TeraPool configurations". We did similar things with the systolic parameters, like the xqueue size. It's always defined in the main config file but only used in systolic configurations.

About the `TERAPOOL` define. We could add something to the makefile that always creates a define in the name of the design/config. So something like this potentially: `vlog_defs += -D$(config)` What do you think?

Comment (yichao-zh) [reply hardware/Makefile]: Hi Sam,

Thanks so much for your suggestions. Now, I have corrected it again. Please help to have a look.
If it looks good for you, I hope we can merge it to the main branch before Xmas!
Please let me know if something still looks bad; thanks a lot for your help!

Cheers,
Yichao.

**Review by SamuelRiedel — APPROVED** [review]
Thanks a lot. LGTM and ready to merge.

**Review by mbertuletti — APPROVED** [review]
All good for me, thank you for fixing the barrier initialization.
Did you modify the Changelog?
